src/libsyntax/attr.rs

```diff
@@ -1103,18 +1116,59 @@ impl IntType {
 
 impl MetaItem {
     fn tokens(&self) -> TokenStream {
-        let ident = TokenTree::Token(self.span, Token::from_ast_ident(self.ident));
-        TokenStream::concat(vec![ident.into(), self.node.tokens(self.span)])
+        let mut idents = vec![];
+        let mut last_pos = BytePos(0 as u32);
+        for (i, segment) in self.ident.segments.iter().enumerate() {
+            let is_first = i == 0;
+            if !is_first {
+                let mod_sep_span = Span::new(last_pos,
+                                             segment.ident.span.lo(),
+                                             segment.ident.span.ctxt());
+                idents.push(TokenTree::Token(mod_sep_span, Token::ModSep).into());
+            }
+            idents.push(TokenTree::Token(segment.ident.span,
+                                         Token::from_ast_ident(segment.ident)).into());
+            last_pos = segment.ident.span.hi();
+        }
+        idents.push(self.node.tokens(self.span));
+        TokenStream::concat(idents)
     }
 
     fn from_tokens<I>(tokens: &mut iter::Peekable<I>) -> Option<MetaItem>
         where I: Iterator<Item = TokenTree>,
     {
-        let (span, ident) = match tokens.next() {
-            Some(TokenTree::Token(span, Token::Ident(ident, _))) => (span, ident),
+        // FIXME: Share code with `parse_path`.
+        let ident = match tokens.next() {
+            Some(TokenTree::Token(span, Token::Ident(mut ident, _))) => {
+                if let Some(TokenTree::Token(_, Token::ModSep)) = tokens.peek() {
+                    ident.span = span;
+                    let mut segments = vec![ast::PathSegment::from_ident(ident)];
+                    tokens.next();
+                    loop {
+                        if let Some(TokenTree::Token(span,
+                                                     Token::Ident(mut ident, _))) = tokens.next() {
+                            ident.span = span;
+                            segments.push(ast::PathSegment::from_ident(ident));
+                        } else {
+                            return None;
+                        }
+                        if let Some(TokenTree::Token(_, Token::ModSep)) = tokens.peek() {
+                            tokens.next();
+                        } else {
+                            break;
+                        }
+                    }
+                    let span = span.with_hi(segments.last().unwrap().ident.span.hi());
+                    ast::Path { span, segments }
+                } else {
+                    ident.span = span;
```

Comment:
This is a hack I would like to change before this PR lands, but I'm not sure what the right thing to do here would be. Apparently `Token::Ident(ident, _)` `ident`s  have the `DUMMY_SP` as their `Span` (Is this intended?). To create the `Path`s correctly we need the correct `Span`, which is the `Span` of `TokenTree::Token(span, _)`.

But the `Path(Segment)::from_ident()` function only takes an `Ident` since the recent changes. Should I add a function `Path(Segment)::from_ident_span()`? Or what would be the right thing to do here?